### ISSUE TITLE
Nodebox facedir no longer rotates textures to follow it.

### ISSUE THREAD

**Issue by VanessaE** [minetest/minetest#8358]
##### Issue type

 - Bug report
 
##### Minetest version

5.0.0 commit 7a0e52acd65e7097977c924f26b387ec3da9baac (fairly close to rc1, I believe)

##### OS / Hardware

Debian

##### Summary

Nodebox-based nodes no longer have their textures rotated correctly.  What looked fine in 0.4.x looks broken in 5.0.0.  For example:

![Screenshot_2019-03-11_12-18-35](https://user-images.githubusercontent.com/1733537/54142787-da9d3200-43fe-11e9-86a4-b8d0c719e1ca.png)

Those green items are Home Decor thin glowlights (https://gitlab.com/VanessaE/homedecor_modpack/blob/master/homedecor/lighting.lua#L52).  The vertically-placed ones on the signpost did not look like this when the structure was built sometime in the middle of 2018, under 0.4.17 (or whatever it was at the time).  Rather, the face you see on top of the light I placed on the ground is what was visible on the camera-facing side of the vertical ones, originally, just as designed.

This affects both old builds and newly-placed items.

The problem is the same whether on a server or in singleplayer.

Other examples:
![Screenshot_2019-03-11_13-18-52](https://user-images.githubusercontent.com/1733537/54143508-5c418f80-4400-11e9-912d-e1c9da272c5e.png)

Noting @pgimeno's comment below, this may only affect hardware-colored nodes, I can't be sure (current glowlights are `colorwallmounted` nodeboxes, the ones in his image are old ones from before that feature came into use for them, instead using simple, baked yellow textures and regular "facedir" mode).

**Comment by pgimeno** [comment]
Can't reproduce with an older homedecor. I haven't tried current.
![glowlight-investigation](https://user-images.githubusercontent.com/4267396/54143476-3630e500-442a-11e9-9c8a-f7c547b198ad.png)

**Comment by pgimeno** [comment]
> \<VanessaE\> that's pre-colorwallmounted

I updated homedecor and that let me reproduce it.

**Comment by pgimeno** [comment]
Bisected to 75320e7e88ab5088a46c9e42c1e789cbdacb13b0

**Comment by VanessaE** [comment]
So yet another setting that should not even exist. Textures cannot be forced "aligned to world" in just about any real-world...world.  Everyone out there is going to have a mixture of nodes that such a global setting will and won't break.

Node defs that don't specify aligned-to-world should have their textures aligned to the node, full stop.

**Comment by paramat** [comment]
I suspect your use of colorwallmounted together with overlays has exposed a bug.

**Comment by jastevenson303** [comment]
Aligning textures to the world is an essential feature in Doom level editing, for what that's worth.

**Comment by VanessaE** [comment]
Perhaps, but this isn't Doom.

**Comment by paramat** [comment]
It's what makes slab and stair nodes look good for brick type nodes, see MTG.

**Comment by VanessaE** [comment]
Then MTG's nodes need to have per-node-def settings, while leaving the global behavior as it was before.

**Comment by jastevenson303** [comment]
http://irc.minetest.net/minetest-hub/2019-03-11#i_5510029

>my backpacks are not a plain nodebox, and they didn't break.  maybe it is just colorwallmounted nodeboxes; definitely worth investigating.

**Comment by pgimeno** [comment]
@numberZero

**Comment by paramat** [comment]
> Then MTG's nodes need to have per-node-def settings, while leaving the global behavior as it was before.

They do. That global setting is confusing because it is default 'enabled' for some reason (can't remember why), but 'align_style' in tile definition has node alignment as the default.
World-aligned textures are opt-in, otherwise a lot more would have broken.

**Comment by VanessaE** [comment]
> World-aligned textures are opt-in, otherwise a lot more would have broken

This would only be the case if that feature were disabled by default.  "Opt in" means the user has to explicitly change something.

**Comment by paramat** [comment]
Well yes, if a server defines a tile as 'align_style = "user"' (which is experimental and rarely used, MTG doesn't use that option) then the clientside setting has effect is 'enabled', maybe it should be 'disabled' instead? But then i don't think a clientside option should exist at all.

**Comment by paramat** [comment]
VanessaE this got me thinking and i opened an issue about something that bothers me #8364 which may be what you are referring to.

**Comment by paramat** [comment]
VanessaE is suggesting `world_aligned_mode = disable` as a short term fix for 5.0.1. It seems this will disable all world-aligned textures, but the feature is new so won't be in much use.

**Comment by paramat** [comment]
It will be good to fix this for 5.0.1 if possible, but i disagree with this bug delaying 5.0.1, it's an aesthetic visual issue that can be somewhat tidied up in the short term by using transparent overlay textures.

**Comment by sfan5** [comment]
this patch makes it work again:
```diff
diff --git a/src/client/mapblock_mesh.cpp b/src/client/mapblock_mesh.cpp
index ed8a073d..6b5ba9f9 100644
--- a/src/client/mapblock_mesh.cpp
+++ b/src/client/mapblock_mesh.cpp
@@ -744,7 +744,7 @@ void getNodeTile(MapNode mn, const v3s16 &p, const v3s16 &dir, MeshMakeData *dat
        u8 dir_i = ((dir.X + 2 * dir.Y + 3 * dir.Z) & 7) * 2;
 
        // Get rotation for things like chests
-       u8 facedir = mn.getFaceDir(ndef);
+       u8 facedir = mn.getFaceDir(ndef, true);
 
        static const u16 dir_to_tile[24 * 16] =
        {
```

however:
1) I don't know what the idea behind the `bool allow_wallmounted = false` param for `getFaceDir` was
(<i>why always return facedir for wallmounted nodes?</i>)
2) Should the same change be applied to `MapblockMeshGenerator::drawMeshNode`? (which is the only other call site)

**Comment by paramat** [comment]
More serious bugs, seems related https://forum.minetest.net/viewtopic.php?f=47&t=22363
Size issue also bisected to https://github.com/minetest/minetest/commit/75320e7e88ab5088a46c9e42c1e789cbdacb13b0

**Comment by sfan5** [comment]
The second issue reported in the thread can be fixed like this:
```diff
diff --git a/src/nodedef.cpp b/src/nodedef.cpp
index 03a163bd..e5aa9f5b 100644
--- a/src/nodedef.cpp
+++ b/src/nodedef.cpp
@@ -872,6 +872,18 @@ void ContentFeatures::updateTextures(ITextureSource *tsrc, IShaderSource *shdsrc
 			recalculateBoundingBox(mesh_ptr[0]);
 			meshmanip->recalculateNormals(mesh_ptr[0], true, false);
 		}
+	} else if ((drawtype == NDT_NODEBOX) &&
+			((node_box.type == NODEBOX_REGULAR) ||
+			(node_box.type == NODEBOX_FIXED)) &&
+			(!node_box.fixed.empty())) {
+		//Convert regular nodebox nodes to meshnodes
+		//Change the drawtype and apply scale
+		drawtype = NDT_MESH;
+		mesh_ptr[0] = convertNodeboxesToMesh(node_box.fixed);
+		v3f scale = v3f(1.0, 1.0, 1.0) * visual_scale;
+		scaleMesh(mesh_ptr[0], scale);
+		recalculateBoundingBox(mesh_ptr[0]);
+		meshmanip->recalculateNormals(mesh_ptr[0], true, false);
 	}
 
 	//Cache 6dfacedir and wallmounted rotated clones of meshes
```

Once again, it's unclear why that block of code was removed.

**Comment by paramat** [comment]
PR fix #8401
